### PR TITLE
support multi websocket connect in one slave

### PR DESCRIPTION
**Pull request by ping-ke**

### CODE REVIEW

File: tests/testdata/debugTests/DataForDebug.json [hunk 1]
```diff
@@ -0,0 +1,101 @@
+{
```
Comment (ninjaahhh) [inline tests/testdata/debugTests/DataForDebug.json:1]: can remove this file? not relevant

File: cluster/config/slave_config.go [hunk 2]
```diff
@@ -40,7 +40,11 @@ func (s *SlaveConfig) UnmarshalJSON(input []byte) error {
 		return err
 	}
 	*s = SlaveConfig(jsonConfig.SlaveConfigAlias)
-	s.WSPort = DefaultWSPort
+	if s.WSPortList == nil {
```
Comment (ninjaahhh) [inline cluster/config/slave_config.go:43]: will an empty slice equal to `nil`?

```
var ls1 []int
ls2  := []int{}

println(ls1 == nil, ls2 == nil)
```

File: cluster/service/node.go [hunk 3]
```diff
@@ -337,33 +338,43 @@ func (n *Node) stopGRPC() {
...
 }
 
 // stopWS terminates the websocket RPC endpoint.
 func (n *Node) stopWS() {
-	if n.wsListener != nil {
-		n.wsListener.Close()
-		n.wsListener = nil
-
-		n.log.Info("WebSocket endpoint closed", "url", fmt.Sprintf("ws://%s", n.config.WSEndpoint))
+	if len(n.wsListenerList) > 0 {
```
Comment (ninjaahhh) [inline cluster/service/node.go:366]: can we iterate an nil array?

```
var ls1 []int
ls2  := []int{}

for i := range ls1 {
  println(i)
}
for i := range ls2 {
  println(i)
}
```

File: cluster/config/slave_config.go (hunk 2 above)
Comment (ping-ke) [reply cluster/config/slave_config.go:43]: change to len(ls) == 0

File: cluster/config/slave_config.go [hunk 4]
```diff
@@ -40,7 +40,11 @@ func (s *SlaveConfig) UnmarshalJSON(input []byte) error {
 		return err
 	}
 	*s = SlaveConfig(jsonConfig.SlaveConfigAlias)
-	s.WSPort = DefaultWSPort
+	if len(s.WSPortList) == 0 {
+		for i, shard := range s.FullShardList {
```
Comment (qizhou) [inline cluster/config/slave_config.go:44]: Need to s.WSPortList = make([].., ...) before setting it?

**Review by qizhou — COMMENTED** [review]
Looks good in general.  Just a few comments.

**Review by ninjaahhh — APPROVED** [review]
looks good to me

**Review by qizhou — APPROVED** [review]

**Pull request merged**